pkg/api/pod/warnings_test.go

```diff
@@ -599,11 +600,435 @@ func TestWarnings(t *testing.T) {
+					},
+				}},
+			}},
+			expected: []string{},
+		},
+		{
+			name: "one container, two ports, different protocols, same port",
+			template: &api.PodTemplateSpec{Spec: api.PodSpec{
+				Containers: []api.Container{{
+					Name: "foo",
+					Ports: []api.ContainerPort{
+						{ContainerPort: 80, Protocol: api.ProtocolUDP},
+						{ContainerPort: 80, Protocol: api.ProtocolTCP},
+					},
+				}},
+			}},
+			expected: []string{},
+		},
+		{
+			name: "one container, two ports, same protocol, same port, different hostport",
+			template: &api.PodTemplateSpec{Spec: api.PodSpec{
+				Containers: []api.Container{{
+					Name: "foo",
+					Ports: []api.ContainerPort{
+						{ContainerPort: 80, Protocol: api.ProtocolTCP, HostPort: 80},
+						{ContainerPort: 80, Protocol: api.ProtocolTCP, HostPort: 81},
+					},
+				}},
+			}},
+			expected: []string{
+				`spec.containers[0].ports[1]: duplicate container port 80/TCP, conflicts with spec.containers[0]`,
+			},
+		},
+		{
+			name: "one container, two ports, same protocol, port and hostPort, different hostIP",
+			template: &api.PodTemplateSpec{Spec: api.PodSpec{
+				Containers: []api.Container{{
+					Name: "foo",
+					Ports: []api.ContainerPort{
+						{ContainerPort: 80, Protocol: api.ProtocolTCP, HostPort: 80, HostIP: "10.0.0.1"},
+						{ContainerPort: 80, Protocol: api.ProtocolTCP, HostPort: 80, HostIP: "10.0.0.2"},
+					},
+				}},
+			}},
+			expected: []string{},
+		},
+		{
+			name: "two containers, one port each, same protocol, different ports",
+			template: &api.PodTemplateSpec{Spec: api.PodSpec{
+				Containers: []api.Container{{
+					Name: "foo",
+					Ports: []api.ContainerPort{
+						{ContainerPort: 80, Protocol: api.ProtocolUDP},
+					},
+				}, {
+					Name: "bar",
+					Ports: []api.ContainerPort{
+						{ContainerPort: 81, Protocol: api.ProtocolUDP},
+					},
+				}},
+			}},
+			expected: []string{},
+		},
+		{
+			name: "two containers, one port each, different protocols, same port",
+			template: &api.PodTemplateSpec{Spec: api.PodSpec{
+				Containers: []api.Container{{
+					Name: "foo",
+					Ports: []api.ContainerPort{
+						{ContainerPort: 80, Protocol: api.ProtocolUDP},
+					},
+				}, {
+					Name: "bar",
+					Ports: []api.ContainerPort{
+						{ContainerPort: 80, Protocol: api.ProtocolTCP},
+					},
+				}},
+			}},
+			expected: []string{},
+		},
+		{
+			name: "two containers, one port each, same protocol, same port, different hostport",
+			template: &api.PodTemplateSpec{Spec: api.PodSpec{
+				Containers: []api.Container{{
+					Name: "foo",
+					Ports: []api.ContainerPort{
+						{ContainerPort: 80, Protocol: api.ProtocolTCP, HostPort: 80},
+					},
+				}, {
+					Name: "bar",
+					Ports: []api.ContainerPort{
+						{ContainerPort: 80, Protocol: api.ProtocolTCP, HostPort: 81},
+					},
+				}},
+			}},
+			expected: []string{
+				`spec.containers[1].ports[0]: duplicate container port 80/TCP, conflicts with spec.containers[0]`,
```

Comment:
```
one container, two ports, same port, same protocol, different hostPort
two containers, one port each, same port, same protocol, different hostPort
```
As hostIP is not set, these two cases will fail.

https://github.com/kubernetes/kubernetes/pull/113245#discussion_r1218639789